### PR TITLE
Update F32 sparse semi-structured support for CUTLASS back-end

### PR DESCRIPTION
**Pull request by alexsamardzic**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116017

cc @nikitaved @pearu @cpuhrsch @amjames @bhosmer @jcaip @ptrblck

### CODE REVIEW

**Comment by alexsamardzic** [comment]
This is follow-up to #115550, finalizing F32 support for sparse semi-structured CUTLASS back-end - it's mainly about converting F32 sparse semi-structured tensor back to dense, that was not critical to have in mentioned PR.

**Comment by alexsamardzic** [comment]
<details>
 <summary>Updated Python/Espresso script to help in finding optimal transformation into sparse semi-structured metadata required by hardware</summary>

```Python
#! /usr/bin/env python

import itertools
import os

os0 = ("0100", "1110")
os1 = ("1100", "1101", "1110")
os2 = ("1000", "1001", "1110")
os3 = ("1110",)
os4 = ("0100", "1001", "1101")
os5 = ("1101",)
os6 = ("1001",)
os7 = ("1001", "1101", "1110")
os8 = ("0100", "1000", "1100")
os9 = ("1100",)
os10 = ("1000",)
os11 = ("1000", "1100", "1110")
os12 = ("0100",)
os13 = ("0100", "1100", "1101")
os14 = ("0100", "1000", "1001")
os15 = ("0100", "1110")


for o0, o1, o2, o3, o4, o5, o6, o7, o8, o9, o10, o11, o12, o13, o14, o15 in itertools.product(os0, os1, os2, os3, os4, os5, os6, os7, os8, os9, os10, os11, os12, os13, os14, os15):
    input = (
        ".i 4\n"
        ".o 4\n"
        f"0000 {o0}\n"
        f"0001 {o1}\n"
        f"0010 {o2}\n"
        f"0011 {o3}\n"
        f"0100 {o4}\n"
        f"0101 {o5}\n"
        f"0110 {o6}\n"
        f"0111 {o7}\n"
        f"1000 {o8}\n"
        f"1001 {o9}\n"
        f"1010 {o10}\n"
        f"1011 {o11}\n"
        f"1100 {o12}\n"
        f"1101 {o13}\n"
        f"1110 {o14}\n"
        f"1111 {o15}\n"
        ".e\n"
    )
    f = open("/tmp/espresso-input.txt", "w")
    f.write(input)
    f.close()
    print("---------------------------------------", flush=True)
    print(input, flush=True)
    print(flush=True)
    os.system("/tmp/espresso </tmp/espresso-input.txt")
    print(flush=True)
</details>

**Comment by alexsamardzic** [comment]
<details>
 <summary>Updated script to benchmark between cuSPARSELt and CUTLASS sparse semi-structured back-ends (also to compare results and benchmar vs. referencedense GEMM)</summary>

```Python
#! /usr/bin/env python

import random

import torch
from torch.testing import make_tensor
import torch.utils.benchmark as benchmark

from torch.sparse.semi_structured import (
    _DTYPE_TO_SEMI_STRUCTURED_SPARSE_CONFIG,
    SparseSemiStructuredTensor,
    to_sparse_semi_structured,
)

SEMI_STRUCTURED_SUPPORTED_DTYPES = _DTYPE_TO_SEMI_STRUCTURED_SPARSE_CONFIG.keys()


def rand_sparse_semi_structured(r, c, dtype, device, choice=None):
    pattern = "2by4" if dtype != torch.float32 else "1by2"
    if pattern == "1by2":
        ksparse = 2
        choices = [[0, 1], [1, 0]]
    elif pattern == "2by4":
        ksparse = 4
        choices = [
            [1, 1, 0, 0],
            [1, 0, 1, 0],
            [1, 0, 0, 1],
            [0, 1, 1, 0],
            [0, 1, 0, 1],
            [0, 0, 1, 1],
        ]
    mask_entries = [choice or random.choice(choices) for i in range(r * c // ksparse)]
    mask = torch.tensor(mask_entries, dtype=torch.bool).view(r, c).to(device)
    dense = make_tensor(r, c, dtype=dtype, device=device)
    dense[dense == 0] = 1  # To prevent zeros except where mask applied.
    dense = dense.masked_fill(~mask, 0)
    return dense


def run_benchmark(m, n, k, device, dtype, rtol, atol, results):
    label = f"dense vs. cuSPARSELt vs. CUTLASS ({dtype})"
    sub_label = f"m:{m:5d} | k:{k:5d} | n:{n:5d}"

    dtypes_out = {
        torch.int8: torch.int32,
        torch.float16: torch.float16,
        torch.bfloat16: torch.bfloat16,
        torch.float32: torch.float32,
    }
    dtype_out = dtypes_out[dtype]

    weight = rand_sparse_semi_structured(m, k, dtype, device)
    input = make_tensor((n, k), dtype=dtype, device=device)

    dtype_dense = torch.float32
    input_dense = input.to(dtype_dense)
    weight_dense = weight.to(dtype_dense)
    output0 = torch.nn.functional.linear(input_dense, weight_dense)

    if dtype.itemsize != 2:
        SparseSemiStructuredTensor._FORCE_CUTLASS = False
        compressed_cusparselt = to_sparse_semi_structured(weight)
        output1 = torch.nn.functional.linear(input, compressed_cusparselt)
        torch.testing.assert_close(output1.to(dtype_dense), output0, rtol=rtol, atol=atol)

    SparseSemiStructuredTensor._FORCE_CUTLASS = True
    compressed_cutlass = to_sparse_semi_structured(weight)
    output2 = torch.nn.functional.linear(input, compressed_cutlass)
    torch.testing.assert_close(output2.to(dtype_dense), output0, rtol=rtol, atol=atol)

    if dtype != torch.int8:
        measurement = benchmark.Timer(
            stmt="linear(input, weight)",
            globals={
                "linear": torch.nn.functional.linear,
                "input": input,
                "weight": weight,
            },
            label=label,
            sub_label=sub_label,
            description="dense latency",
        ).blocked_autorange()
        results.append(measurement)

    if dtype.itemsize != 2:
        measurement = benchmark.Timer(
            stmt="linear(input, weight)",
            globals={
                "linear": torch.nn.functional.linear,
                "input": input,
                "weight": compressed_cusparselt,
            },
            label=label,
            sub_label=sub_label,
            description="cuSPARSELt latency",
        ).blocked_autorange()
        results.append(measurement)

    measurement = benchmark.Timer(
        stmt="linear(input, weight)",
        globals={
            "linear": torch.nn.functional.linear,
            "input": input,
            "weight": compressed_cutlass,
        },
        label=label,
        sub_label=sub_label,
        description="CUTLASS latency",
    ).blocked_autorange()
    results.append(measurement)


if __name__ == "__main__":
    device = "cuda"

    shapes = [
        # distilbert shapes
        (768, 3072, 768),
        (3072, 768, 3072),
        # jiecao shapes
        (1024, 1536, 2048),
        (1024, 9408, 2048),
        (1024, 3200, 2048),
        (1024, 256, 9472),
        (1024, 10240, 256),
        (1024, 256, 12608),
        (1024, 2560, 1024),
        (1024, 512, 10240),
        (1024, 10240, 512),
        (1024, 2048, 1024),
        (1024, 512, 512),
        (1024, 1024, 1024),
        (1024, 2048, 2048),
        (2048, 1536, 2048),
        (2048, 9408, 2048),
        (2048, 3200, 2048),
        (2048, 256, 9472),
        (2048, 10240, 256),
        (2048, 256, 12608),
        (2048, 2560, 1024),
        (2048, 512, 10240),
        (2048, 10240, 512),
        (2048, 2048, 1024),
        (2048, 512, 512),
        (2048, 1024, 1024),
        (2048, 2048, 2048),
    ]

    torch.set_printoptions(
        precision=3,
        threshold=None,
        edgeitems=4,
        linewidth=460,
        profile=None,
        sci_mode=False,
    )

    for dtype in SEMI_STRUCTURED_SUPPORTED_DTYPES:
        results = []

        if dtype == torch.float32:
            # Inputs are converted to TF32 internally for sparse GEMM,
            # so make dense GEMM to do the same for matching results.
            orig = torch.backends.cuda.matmul.allow_tf32
            torch.backends.cuda.matmul.allow_tf32 = True

        rtol, atol = 1e-3, 1e-3
        if dtype == torch.float16 or dtype == torch.bfloat16:
            rtol, atol = 5e-3, 5e-2
        if dtype == torch.float32:
            rtol, atol = 1e-3, 5e-1

        for m, k, n in shapes:
            try:
                print(
                    f"m = {m:5d}, n = {n:5d}, k = {k:5d}, dtype = {dtype} ... ",
                    end="",
                )
                run_benchmark(m, n, k, device, dtype, rtol, atol, results)
                print("ok")
            except Exception:
                print("failed")
                continue

        if dtype == torch.float32:
            torch.backends.cuda.matmul.allow_tf32 = orig

        compare = benchmark.Compare(results)
        compare.colorize(rowwise=True)
        compare.print()
```
</details>

File: torch/sparse/_semi_structured_conversions.py [hunk 1]
```diff
@@ -1,7 +1,50 @@
...
+# sparse GEMM executed on tensor cores, this is layout described by
+# ColumnMajorInterleaved<2> data structure, in
+# include/cutlass/layout/matrix.h of CUTLASS source tree.  The
+# reordering of meta matrix into meta_reordered matrix calculated
+# according to these segments of CUTLASS code is re-implemented here.
+# Note that this calculation produces offsets for scattering metadata
+# matrix elements into reordered metadata matrix elements (or,
+# equivalently, for gathering reordered metadata matrix element back
+# into metadata matrix elements).
+def _calculate_meta_reordering_scatter_offsets(m, meta_ncols, meta_dtype, device):
```
Comment (jcaip) [inline torch/sparse/_semi_structured_conversions.py:17]: Does the code in this function and the code [here](https://github.com/pytorch/pytorch/pull/116017/files#diff-c7ba2441e5192bccd1becb71844c84d6657506c4abdf2300c3f5b829aad1c552L179-L216) do the same thing - if so is there a reason why we no longer need the "fast" method?

**Review by jcaip — APPROVED** [review]
lgtm

File: torch/sparse/_semi_structured_conversions.py (hunk 1 above)
Comment (alexsamardzic) [reply torch/sparse/_semi_structured_conversions.py:17]: Yes, I replaced this complicated code with a direct rewrite (that I actually started from) of CPU code provided in the CUTLASS library doing the conversion.  I found the code related to conversion back to dense tensor time consuming to extend for float32.  The original code was written with the idea that it will be always compiled for Triton, even in eager mode; as we moved away from that approach, my opinion is that it's going to be easier to maintain and eventually extend this simpler code.  Of course, we can always revert; on the other side, I hope CUTLASS guys may provide GPU implementation themselves, in particular as you mentioned earlier that it's going to be there in cuSPARSELt.

Comment (alexsamardzic) [reply torch/sparse/_semi_structured_conversions.py:17]: Question (unrelated): Do we have a set of shapes that would be relevant to run the benchmark script attached above?  The shapes cited in the script are ones that I'm using for while for this purpose, but it would be interesting to benchmark on larger ones.

Comment (jcaip) [reply torch/sparse/_semi_structured_conversions.py:17]: @alexsamardzic That makes sense, I may try to spend a day to write some fast cuSPARSELt torch.compile routines, but long term we're looking to merge in the fast ones @danthe3rd wrote. No need to revert, just wanted to understand the code. 

> Question (unrelated): Do we have a set of shapes that would be relevant to run the benchmark script attached above?

Let me collect this together for all the relevant models (SAM, stable-diffusion, LLaMa) and update the script

**Comment by alexsamardzic** [comment]
Latest update is just about slightly increasing tolerance for `test_linear_cutlass` test for float32 data-type, in order to silence [issue](https://github.com/pytorch/pytorch/issues/116045) reported by CI.
